Additions for libre-audit & postgres (#96)
libremfg/libremfg.github.io: content/deploy/maintain/audit.md

---
title: 'Maintain Audit PostgreSQL'
date: '2024-03-26T11:20:56-03:00'
categories: ["how-to"]
description: How to maintain Audit PostgreSQL on your Rhize deployment
weight: 300
menu:
  main:
    parent: maintain
    identifier:
---

> This guide shows you the procedure to maintain your Audit Postgres database on your Rhize Kubernetes deployment.

Reviewer: I think we can be more specific than "Maintain": what is the actual action being documented? That's what we should title it.

> ---
> title: 'Maintain Audit PostgreSQL'
> date: '2024-03-26T11:20:56-03:00'
> categories: ["how-to"]
> description: How to maintain Audit PostgreSQL on your Rhize deployment
> weight: 300
> menu:
>   main:
>     parent: maintain
>     identifier:
> ---
> 
> This guide shows you the procedure to maintain your Audit Postgres database on your Rhize Kubernetes deployment.

Reviewer: ```suggestion
---
title: 'Archive the PostgreSQL Audit trail'
date: '2024-03-26T11:20:56-03:00'
categories: ["how-to"]
description: How to archive a partition of the Audit trail on your Rhize deployment
weight: 300
menu:
  main:
    parent: maintain
    identifier:
---

The [audit trial]({{< relref "/deploy/audit" >}}) can generate a high volume of data, so it is a good practice to periodically _archive_ portions of it. An archive separates a portion of the data from the database and keeps it for long-term storage. This process involves the use of PostgresQL [Table Partitions](https://www.postgresql.org/docs/current/ddl-partitioning.html).
```

## Prerequisites

Before you start, ensure you have the following:

> - A designated backup location, for example `~/rhize-backups/libre-audit`.

Reviewer: Would this not conflict with the backups in /deploy/backup/audit? Maybe we should give it a different title

> - A designated backup location, for example `~/rhize-backups/libre-audit`.

Reviewer: ```suggestion
- A designated backup location, for example `~/rhize-backups/libre-audit-archive`.
```

> The [audit trial]({{< relref "/how-to/audit" >}}) can generate a high volume of data, so it is a good practice to periodically _archive_ portions of it. An archive separates a portion of the data from the database and keeps it for long-term storage. This process involves the use of PostgresQL [Table Partitions](https://www.postgresql.org/docs/current/ddl-partitioning.html).

Reviewer 2: Doing this speeds up the query of existing data, whilst allowing for cost effective storage of older data.

- Access to the [Rhize Kubernetes Environment](/deploy/install/setup-kubernetes)
{{% param pre_reqs %}}

Also, before you start, confirm you are in the right context and namespace.

{{% param k8s_cluster_ns %}}

## Steps

> Maintaining Audit PostgreSQL involves three steps:

Reviewer: ```suggestion
To archive the PostgreSQL Audit trail, follow these steps:
```

> 

Reviewer 2: Two new lines seem out of place

1. Detach target partitions from main table:

    ```bash
    kubectl exec -i audit-postgresql-0 -- psql -h localhost \
      -d audit -U <DB_USER> \
      -c 'alter table audit_log detach partition audit_log_p20240101;'
    ```

> 1. Backup partition table:

Reviewer: ```suggestion
1. Backup the partition table:
```

> 

Reviewer 2: Before detatching / dumping, provide instruction on how to list the current partitions:

1. Based on your retention period query for the names of the existing partitions:

   ```bash
   kubectl exec -i audit-postgresql-0 -- psql -h localhost \ 
     -d audit -U <DB_USER> \
     -c 'SELECT * FROM x'
   ```
   
    Record the name `<PARTITION_NAME>` of the partition you wish to detach and archive.

>       -c 'alter table audit_log detach partition audit_log_p20240101;'

Reviewer 2: replace audit_log_p20240101 with `<PARTITION_NAME>`

    ```bash
    pg_dump -U <DB_USER> -h audit-postgres-0 -p5433 \
      --file ./audit-p20240101.sql --table public.audit_log_p20240101 audit
    ```

   On success, the backup creates a GZIP file, `audit-p20240101.sql`.
   To check that the backup succeeded, unzip the files and inspect the data.

> 1. Drop the partition table to remove from database:

Reviewer: ```suggestion
1. Drop the partition table to remove from the database:
```

>       --file ./audit-p20240101.sql --table public.audit_log_p20240101 audit

Reviewer 2: Replace `audit_log_p20240101` with `<PARTITION_NAME>`
